tests/detail/numeric_util_test_suite.cpp

```diff
@@ -219,5 +221,18 @@ class numeric_test_suite : public test_suite
 
         static_assert(xlnt::detail::abs(-1.23) == 1.23, "constexpr");
     }
+
+    void test_locale_comma ()
+    {
+        struct SetLocale
+        {
+            SetLocale() {xlnt_assert(setlocale(LC_ALL, "de_DE") != nullptr);} // If failed, please install de_DE locale to correctly run this test.
```

Comment:
I do understand that this is necessary (for now), but I'm not exactly happy with the fact that this requires the German (Germany) locale to be installed on the computer. Furthermore, changing the global locale is always something risky to do. However, to avoid all of this, we would need to write a parser that is more robust (in other words, a wrapper around std::strtod) - but this would be more work and way beyond the scope of this Pull Request.

I see it this way: this Pull Request fixes a really nasty bug and even provides a unit test to prevent such issues to happen in the future. There are more things that could be improved, but I think this PR should be accepted as it is (for now).

Reply:
Even if we write/use a more robust parser, it may be still a good idea to have some unit tests for non-default locales.

If you want, we could add an option to disable all unit tests using non-default locales, but I'm not sure it is a big deal to require the user to install the German locale to run the unit tests.
Alternatively, we could remove changing the locale in the unit test and setup a CircleCI or AppVeyor job using the de_DE locale for all unit tests.

Reply:
I think it's a very good idea to test both decimal separators, yes. However, what I have in mind is a wrapper that does the following:
1. Wrap the C function `strtod` in a function that does the same as std::stod (note: its C++ counterpart) but does NOT throw exceptions (easier to handle and does not slow down parsing unnecessarily). I think the C++ parsing functions get a lot right, except that they require the programmer to do exception handling everywhere which is annoying and might cause performance issues.
2. Try to parse the number using `strtod` as it is, and check the end of the parsed string (second parameter of `strtod`). If a potential decimal separator (more about that below) was detected, the replacements must be made.
3. If the parsing stopped at the previous step, get the decimal separator of the current locale (more about that below), and if it stopped at the `.` (dot), replace it with the decimal separator of the current locale. Reparse the replaced string.
4. If the parsing stopped (again) at the previous step, and it stopped at the `,` (comma), replace it with the decimal separator of the current locale. Reparse the replaced string.
5. If all of the steps above have failed, the number cannot be parsed.

I used this technique in other projects at places where the input can literally be in any locale, but still needs to be parsable in all cases. The other advantage is that this technique does the minimum number of replacements and does not require to do replacements in the whole (potentially longer) string - although this is probably not a use case for XLSX files.

As for the "current locale", we could call `std::locale("")` and cache it in a static variable to avoid constructing it every time we need it. Then we can pass it as a `const xlnt::optional<std::locale> & = {}` to the parsing function. In case the `xlnt::optional` is empty, it will use the cached locale, otherwise it will use the provided locale (which is super useful for unit tests).

I know the technique I explained above might seem to be way too complicated, but I have experience with parsing random data from unknown sources, and based on that experience I can say that NOT having a robust parser will create tons of issues down the road :sweat_smile:

Furthermore, due to working with multi-threaded software, I try to avoid using C functions that have a global state as much as possible where there's an alternative, e.g. using `std::locale` from C++ rather than `setlocale` and `localeconv` from C. We need to keep in mind that XLNT is not multi-threaded and `setlocale` is only called for unit tests at the moment, so (at least currently) this may be a matter of preference - but I prefer not using C locale functions if possible to avoid global state.

Reply:
I'm happy to accept a PR implementing this robust parser.
I agree that avoiding `setlocale` call (in the future) is a good idea, if we could still write a unit test for a non-default locale (e.g. by providing your optional locale argument), but I assume this would still require the locale to be installed on the system to run the unit test successfully.

Reply:
I did not test the following code, but something like this should work:
```cpp
// Define this somewhere
class custom_numpunct : public std::numpunct<char>
{
protected:
    char do_decimal_point() const override { return ','; }
};

// Then take either the system "" or the "C" locale (in this case it won't matter) and replace it with our custom numpunct.
std::locale loc(std::locale(""), new custom_numpunct());
```

As for using the `.` (dot) as decimal separator, we can simply use `std::locale::classic()` for that purpose. All of this will then work without contributors or the CI having to install an additional locale  :slightly_smiling_face: